### ISSUE TITLE
Parser to return UTF-8

### ISSUE THREAD

**Issue by SunMar** [php-edifact/edifact#140]
Hi,

Thanks for the great package! It makes working with EDIFACT files much easier.

We however recently ran into an issue with encodings and were thinking maybe it's nice if it's possible for the parser to indicate that we always want the data returned as `UTF-8`. Then depending on `$this->syntaxID` the parser knows what the `from` encoding is to use (assuming the file is coming in the correct encoding as per the EDIFACT specifications).

Our idea was to maybe add a parameter in the `get()` method like `get($encoding = null)`. It will then remain backwards compatible and not change the encoding if `get()` is called without any argument, but if you want `UTF-8` returned you could call `$parser->get('UTF-8')` and then based on `$this->syntaxID` and $this->charsets` it can convert all the data from `ASCII` or `ISO-8859-*` to `UTF-8` using `mb_convert_encoding()`.

Happy to make a PR for this, but wanted to verify the preferred approach before doing so and make sure we're not overlooking any potential issues with this change.

After getting the data we use `EDI\Interpreter->prepare()` to get our data array, which is in the end what we want to fully contain `UTF-8`.

**Comment by sabas** [comment]
Can you elaborate on the issue? The syntax checking does change only the stripping regular expression, but you can override it with setstripregex...
If you have implemented already the solution you are proposing we can discuss it no problem :)

**Comment by SunMar** [comment]
@sabas Well after the data is parsed, we want to put it in a database, but the text in the database needs to be in UTF-8. Right now the library does not do any encoding conversions, meaning if it's `UNOC` the data will be in `ISO-8859-1`, if it's `UNOD` it will be `ISO-8859-2`, etc. This is fine for `ASCII` characters, but any characters from the extended codes (like `é`, `ö`, etc.) will get mangled if we try to insert this directly into the database or show it in the UI without converting it first to `UTF-8`.  Since the encoding of the data depends on the `syntaxID`, when we get the data back from the parser we don't know which encoding to convert from.

Our workaround right now is to copy/paste the `$this->charsets` array in our own code, and then do the `mb_convert_encoding()` there, but we figured it might make sense if the library has an option for this.

Something like below, for illustration purposes, not actual implementation :). With an option to specify the output encoding we can get data in a uniform encoding, regardless of the encoding of the input data, we can then safely store it in the database or show it in the UI.

```php
public function get(?string $encoding = null): array
{
    if (empty($this->parsedfile)) {
        $this->parse();
    }

    if (null === $encoding) {
        return $this->parsedfile;
    }

    return $this->convertEncoding($this->parsedfile, self::$charsets[$this->syntaxID], $encoding);
}

private function convertEncoding($data, string $from, string $to)
{
    if (is_array($data)) {
        foreach ($data as $k => $v) {
            $data[$k] = $this->convertEncoding($v, $from, $to);
        }
    } elseif (is_string($data)) {
        $data = mb_convert_encoding($data, $to, $from);
    }

    return $data;
}
```

Hopefully that explains it. If not happy to answer any other questions :).

**Comment by sabas** [comment]
Ok sounds good! Although I wouldn't rely too much on what syntaxID is provided in the message, as I saw extended charsets used in messages stated as UNOA :)

**Comment by SunMar** [comment]
@sabas Yes, we've also had documents that don't match the encoding that it's supposed to be, like receiving `UNOC` documents as `UTF-8`. We'll see if we can do something a little generic.

**Comment by sabas** [comment]
Solved by #141

**Comment by gaxweb** [comment]
I mean, this is fine I guess. But you could have just done that yourself after running the get() method.

I'm also against messing with the encoding of the provided EDI file. I think it leads down the road of encoding issues. Any kind of conversion should be out of scope for this project. Everyone should fix whatever there is to fix until the file adheres to the spec it supposedly represents. That's why I think the `setSourceEncoding` method is ill advised. The parser should just reject files until they're valid.
